### PR TITLE
Fix the behavior of `jax.scipy.stats.sem` when `keepdims=True`.

### PR DESCRIPTION
**Pull request by rajasekharporeddy**
`jax.scipy.stats.sem` does not follow the behavior of `scipy.stats.sem` when`keepdims=True`. This PR fixes this.

Current behavior:
```python
>>> import jax, scipy
>>> 
>>> x = jnp.array([[1, 2, 3],
...                [4, 5, 6],
...                [7, 8, 9]])
>>> scipy.stats.sem(x, axis=1, keepdims=True)
array([[0.57735027],
       [0.57735027],
       [0.57735027]])
>>> jax.scipy.stats.sem(x, axis=1, keepdims=True)
Array([0.57735026, 0.57735026, 0.57735026], dtype=float32)
```
With this PR:
```python
>>> jax.scipy.stats.sem(x, axis=1, keepdims=True)
Array([[0.57735026],
       [0.57735026],
       [0.57735026]], dtype=float32)
```

### CODE REVIEW

File: jax/_src/scipy/stats/_core.py [hunk 1]
```diff
@@ -203,13 +203,28 @@ def sem(a: ArrayLike, axis: int | None = 0, ddof: int = 1, nan_policy: str = "pr
...
+  if keepdims:
+    if axis is None:
+      output_shape = tuple(1 for i in input_shape)
+    else:
+      output_shape = tuple(1 if i == axis else s for i, s in enumerate(input_shape))
+  else:
+    if axis is None:
+      output_shape = ()
+    else:
+      output_shape = tuple(s for i, s in enumerate(input_shape) if i != axis)
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:217]: better I think would be to pass `keepdims` to `std`, `nanstd`, and `sum` below rather than adding all this extra shape logic here. What do you think?

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/scipy/stats/_core.py (hunk 1 above)
Comment (rajasekharporeddy) [reply jax/_src/scipy/stats/_core.py:217]: Yes, agree with you. But when I use it, `scipy_stats_test.py` fails the some tests.

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:217]: Then I'd figure out why it's failing and fix it 😁

Comment (rajasekharporeddy) [reply jax/_src/scipy/stats/_core.py:217]: It fails for `axis=None` cases, especially.

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:217]: If you'd like me to help debug the failures and suggest solutions, then you could push what you have to this branch so I can see it.

Comment (rajasekharporeddy) [reply jax/_src/scipy/stats/_core.py:217]: The current changes with extra shape logic passes all the tests. Even if I test with the following command.
```
JAX_NUM_GENERATED_CASES=50 pytest -n auto tests/scipy_stats_test.py -k testSEM
```

If I try passing `keepdims` to `std`, `nanstd`, and `sum`, some tests are failing.

I am pushing the commit with the changes suggested by you here. Please help me debug the failures and suggest solution. Thanks!

File: jax/_src/scipy/stats/_core.py [hunk 2]
```diff
@@ -207,9 +207,9 @@ def sem(a: ArrayLike, axis: int | None = 0, ddof: int = 1, nan_policy: str = "pr
     b = b.ravel()
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:207]: You need to remove this block if you want to push this logic into the reductions below here.

File: jax/_src/scipy/stats/_core.py [hunk 3]
```diff
@@ -207,9 +207,9 @@ def sem(a: ArrayLike, axis: int | None = 0, ddof: int = 1, nan_policy: str = "pr
...
     axis = 0
   if nan_policy == "propagate":
-    return b.std(axis, ddof=ddof) / jnp.sqrt(b.shape[axis]).astype(b.dtype)
+    return b.std(axis, ddof=ddof, keepdims=keepdims) / jnp.sqrt(b.shape[axis]).astype(b.dtype)
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:210]: When `axis=None`, `b.shape[axis]` will have to be `b.size` instead.

File: jax/_src/scipy/stats/_core.py (hunk 2 above)
Comment (rajasekharporeddy) [reply jax/_src/scipy/stats/_core.py]: Thanks! Removed.

File: jax/_src/scipy/stats/_core.py (hunk 3 above)
Comment (rajasekharporeddy) [reply jax/_src/scipy/stats/_core.py]: Thanks! Modified accordingly.

File: jax/_src/scipy/stats/_core.py [hunk 4]
```diff
@@ -203,13 +203,13 @@ def sem(a: ArrayLike, axis: int | None = 0, ddof: int = 1, nan_policy: str = "pr
     array
   """
   b, = promote_args_inexact("sem", a)
-  if axis is None:
-    b = b.ravel()
-    axis = 0
   if nan_policy == "propagate":
-    return b.std(axis, ddof=ddof) / jnp.sqrt(b.shape[axis]).astype(b.dtype)
```
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:210]: Since we're only changing one term, we should avoid duplicating the rest of the expression. Perhaps something like this?
```python
size = b.size if axis is None else b.shape[axis]
return b.std(axis, ddof=ddof, keepdims=keepdims) / jnp.sqrt(size).astype(b.dtype)
```

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/scipy/stats/_core.py (hunk 4 above)
Comment (rajasekharporeddy) [reply jax/_src/scipy/stats/_core.py:210]: Done

**Review by jakevdp — COMMENTED** [review]
Looks good! If you can squash into a single commit, we'll get this merged. Thanks!

**Comment by rajasekharporeddy** [comment]
Squashed into single commit. Thank you for the help @jakevdp.

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by jakevdp** [comment]
It looks like we'll have to test `keepdims` differently for older Scipy versions, which don't support `keepdims` in `sem`.

**Comment by rajasekharporeddy** [comment]
Could you please suggest me how I can add the tests that will be compatible to older scipy versions.

File: tests/scipy_stats_test.py [hunk 5]
```diff
@@ -1633,21 +1633,24 @@ def testRankData(self, shape, dtype, axis, method):
...
 
     rng = jtu.rand_default(self.rng())
     args_maker = lambda: [rng(shape, dtype)]
 
-    scipy_fun = partial(osp_stats.sem, axis=axis, ddof=ddof, nan_policy=nan_policy)
-    lax_fun = partial(lsp_stats.sem, axis=axis, ddof=ddof, nan_policy=nan_policy)
+    scipy_fun = partial(osp_stats.sem, axis=axis, ddof=ddof, nan_policy=nan_policy,
+                        keepdims=keepdims)
+    lax_fun = partial(lsp_stats.sem, axis=axis, ddof=ddof, nan_policy=nan_policy,
+                      keepdims=keepdims)
```
Comment (jakevdp) [inline tests/scipy_stats_test.py:1653]:
```suggestion
    kwds = {} if scipy_version < (1, 11) else {'keepdims': keepdims}
    scipy_fun = partial(osp_stats.sem, axis=axis, ddof=ddof, nan_policy=nan_policy, **kwds)
    lax_fun = partial(lsp_stats.sem, axis=axis, ddof=ddof, nan_policy=nan_policy, **kwds)
```

Comment (rajasekharporeddy) [reply tests/scipy_stats_test.py]: Updated the test

**Review by jakevdp — APPROVED** [review]
